plugin/pkg/admission/runtimeclass/admission.go

```diff
@@ -0,0 +1,199 @@
+	runtimeClassInformer := f.Node().V1beta1().RuntimeClasses()
+	r.SetReadyFunc(runtimeClassInformer.Informer().HasSynced)
+	r.lister = runtimeClassInformer.Lister()
+}
+
+func (r *RuntimeClass) SetExternalKubeClientSet(client kubernetes.Interface) {
+	r.client = client
+}
+
+func (r *RuntimeClass) ValidateInitialization() error {
+	if r.lister == nil {
+		return fmt.Errorf("missing RuntimeClass lister")
+	}
+	if r.client == nil {
+		return fmt.Errorf("missing client")
+	}
+	return nil
+}
+
+// Admit makes an admission decision based on the request attributes
+func (r *RuntimeClass) Admit(attributes admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	return admissionAction(r, attributes, "admit")
+}
+
+// Validate makes sure that pod adhere's to RuntimeClass's definition
+func (r *RuntimeClass) Validate(attributes admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	return admissionAction(r, attributes, "validate")
+}
+
+// NewRuntimeClass creates a new always pull images admission control handler
+func NewRuntimeClass() *RuntimeClass {
+	return &RuntimeClass{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}
+}
+
+// admissionAction handles Admit and Validate phases of admission, switching based on the admissionPhase parameter
+func admissionAction(r *RuntimeClass, attributes admission.Attributes, admissionPhase string) (err error) {
+
+	// Ignore all calls to subresources or resources other than pods.
+	if shouldIgnore(attributes) {
+		return nil
+	}
+	pod, ok := attributes.GetObject().(*api.Pod)
+	if !ok {
+		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+
+	// Get RuntimeClass. Not finding a given runtimeClass does not return an error. We will want to continue in the
+	// case of Validate phase, even if a RuntimeClass is not defined
+	runtimeClass, err := getRuntimeClass(r, pod, pod.Spec.RuntimeClassName)
+	if err != nil {
+		return err
+	}
+
+	switch admissionPhase {
+	case "admit":
+		err := setOverhead(pod, runtimeClass)
+		if err != nil {
+			return err
+		}
+
+	case "validate":
+		err := validateOverhead(pod, runtimeClass)
+		if err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("invalid value for admissionPhase: %s", admissionPhase)
+	}
+
+	return nil
+}
+
+// getRuntimeClass will return a reference to the RuntimeClass object if it is found. If it cannot be found, or a RuntimeClassName
+// is not provided in the pod spec, *node.RuntimeClass returned will be nil
+func getRuntimeClass(r *RuntimeClass, pod *api.Pod, runtimeClassName *string) (*node.RuntimeClass, error) {
+
+	runtimeClass := &node.RuntimeClass{}
+
+	if runtimeClassName != nil {
+		v1beta1runtimeClass, _ := r.lister.Get(*runtimeClassName)
+		if v1beta1runtimeClass != nil {
+			err := nodev1beta1.Convert_v1beta1_RuntimeClass_To_node_RuntimeClass(v1beta1runtimeClass, runtimeClass, nil)
```

Comment:
what's the point of conversion here? all the resources except for `pods` are expected to be external types under `k8s.io/api/...` pkg.

Reply:
Sorry if it isn't totally clear, but I am ultimately assigning the Overhead value to the particular PodSpec, so need the internal type. The (v1beta1) lister returns v1beta1 types, so need to convert for application to the pod.

Reply:
I'd prefer to keep using the external type, and only do the conversion when we go to actually copy the overhead over.